File: examples/javascript/simple_crud.js
/**
 * Simple CRUD Operations using Direct HTTP Requests
 *
 * Demonstrates basic Create, Read, Update, Delete operations
 * using raw HTTP/fetch API - no client library required
 */

const dotenv = require("dotenv");

dotenv.config();

const BASE_URL = process.env.API_BASE_URL || "http://localhost:8080";
const API_KEY = process.env.API_BASE_KEY || "a-test-api-key-from-ekodb";

let authToken = null;

async function getAuthToken() {
  if (authToken) return authToken;

  const response = await fetch(`${BASE_URL}/api/auth/token`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ api_key: API_KEY }),
  });

  if (!response.ok) {
    throw new Error(`Auth failed: ${response.status}`);
  }

  const data = await response.json();
  authToken = data.token;
  return authToken;
}

async function request(method, path, body = null) {
  const token = await getAuthToken();
  const options = {
    method,
    headers: {
      Authorization: `Bearer ${token}`,
      "Content-Type": "application/json",
    },
  };

  if (body) {
    options.body = JSON.stringify(body);
  }

  const response = await fetch(`${BASE_URL}${path}`, options);

  if (!response.ok && response.status !== 404) {
    throw new Error(`HTTP ${response.status}: ${await response.text()}`);
  }

  if (response.status === 404) {
    return null;
  }

  return response.json();
}

async function main() {
  console.log("=== Simple CRUD Operations (Direct HTTP) ===\n");

  await getAuthToken();
  console.log("✓ Authentication successful");

  // Example 1: Insert a document
  console.log("\n=== Insert Document ===");
  const doc = await request("POST", "/api/insert/test_collection", {
    name: "Test Record",
    value: 42,
    active: true,
  });
  console.log("Inserted:", doc);
  const docId = doc.id;

  // Example 2: Find by ID
  console.log("\n=== Find by ID ===");
  const foundDoc = await request("GET", `/api/find/test_collection/${docId}`);
  console.log("Found:", foundDoc);

  // Example 3: Find with query
  console.log("\n=== Find with Query ===");
  const docs = await request("POST", "/api/find/test_collection", {
    filter: {
      type: "Condition",
      content: {
        field: "active",
        operator: "Eq",
        value: true,
      },
    },
    limit: 10,
  });
  console.log("Found documents:", docs);

  // Example 4: Update document
  console.log("\n=== Update Document ===");
  const updated = await request("PUT", `/api/update/test_collection/${docId}`, {
    name: "Updated Record",
    value: 100,
  });
  console.log("Updated:", updated);

  // Example 5: Delete document
  console.log("\n=== Delete Document ===");
  await request("DELETE", `/api/delete/test_collection/${docId}`);
  console.log("Deleted document");

  console.log("\n✓ All CRUD operations completed successfully");
}

main().catch((error) => {
  console.error("Error:", error);
  process.exit(1);
});

File: examples/python/simple_crud.py
#!/usr/bin/env python3
"""
Simple CRUD Operations using Direct HTTP Requests

Demonstrates basic Create, Read, Update, Delete operations
using raw HTTP requests - no client library required
"""

import os

import requests
from dotenv import load_dotenv

load_dotenv()

BASE_URL = os.getenv("API_BASE_URL", "http://localhost:8080")
API_KEY = os.getenv("API_BASE_KEY", "a-test-api-key-from-ekodb")

auth_token = None
COLLECTION = "simple_crud_example_py"


def get_auth_token():
    global auth_token
    if auth_token:
        return auth_token

    response = requests.post(
        f"{BASE_URL}/api/auth/token",
        json={"api_key": API_KEY},
        headers={"Content-Type": "application/json"},
    )
    response.raise_for_status()
    auth_token = response.json()["token"]
    return auth_token


def request(method, path, body=None):
    token = get_auth_token()
    headers = {
        "Authorization": f"Bearer {token}",
        "Content-Type": "application/json",
    }

    response = requests.request(method, f"{BASE_URL}{path}", json=body, headers=headers)

    if response.status_code == 404:
        return None

    if not response.ok:
        raise Exception(f"HTTP {response.status_code}: {response.text}")

    if not response.text:
        return {}

    return response.json()


def run_examples():
    print("=== Simple CRUD Operations (Direct HTTP) ===\n")

    get_auth_token()
    print("✓ Authentication successful")

    # Example 1: Insert a document
    print("\n=== Insert Document ===")
    doc = request(
        "POST",
        f"/api/insert/{COLLECTION}",
        {
            "name": "Test Record",
            "value": 42,
            "active": True,
        },
    )
    print(f"Inserted: {doc}")
    doc_id = doc["id"]

    # Example 2: Find by ID
    print("\n=== Find by ID ===")
    found_doc = request("GET", f"/api/find/{COLLECTION}/{doc_id}")
    print(f"Found: {found_doc}")

    # Example 3: Find with query
    print("\n=== Find with Query ===")
    docs = request(
        "POST",
        f"/api/find/{COLLECTION}",
        {
            "filter": {
                "type": "Condition",
                "content": {
                    "field": "active",
                    "operator": "Eq",
                    "value": True,
                },
            },
            "limit": 10,
        },
    )
    print(f"Found documents: {docs}")

    # Example 4: Update document
    print("\n=== Update Document ===")
    updated = request(
        "PUT",
        f"/api/update/{COLLECTION}/{doc_id}",
        {
            "name": "Updated Record",
            "value": 100,
        },
    )
    print(f"Updated: {updated}")

    # Example 5: Delete document
    print("\n=== Delete Document ===")
    request("DELETE", f"/api/delete/{COLLECTION}/{doc_id}")
    print("Deleted document")


def main():
    request("DELETE", f"/api/collections/{COLLECTION}")
    operation_error = None
    try:
        run_examples()
    except BaseException as error:  # noqa: BLE001 - cleanup must run on cancellation
        operation_error = error

    try:
        request("DELETE", f"/api/collections/{COLLECTION}")
    except Exception as cleanup_error:
        if operation_error is not None:
            operation_error.add_note(f"cleanup also failed: {cleanup_error}")
        else:
            raise
    if operation_error is not None:
        raise operation_error

    print("\n✓ All CRUD operations completed successfully")


if __name__ == "__main__":
    main()
